File: apps/backend/routers/credit_committee.py
from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import Response
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from core.access_control import (
    COMPANY_MANAGEMENT_ROLES,
    LENDING_ROLES,
    TenantContext,
    get_user_context,
    require_tenant_roles,
)
from database.models.credit_committee import CreditCommitteeCase
from database.models.enums import UserRole
from database.session import get_db
from services.credit_committee_report_service import build_credit_committee_pdf
from services.credit_committee_service import (
    application_or_404,
    assessment_payload,
    case_or_404,
    case_payload,
    cast_vote,
    condition_payload,
    configure_case_governance,
    dashboard_payload,
    ensure_case,
    finalize_case,
    submit_assessment,
    update_condition,
)


router = APIRouter(prefix="/credit-committee", tags=["Credit Committee & Underwriting"])

VIEW_ROLES = COMPANY_MANAGEMENT_ROLES | LENDING_ROLES | {
    UserRole.RISK_MANAGER,
    UserRole.COMPLIANCE_OFFICER,
    UserRole.AUDITOR,
}
ANALYST_ROLES = COMPANY_MANAGEMENT_ROLES | {
    UserRole.BRANCH_MANAGER,
    UserRole.CREDIT_ANALYST,
    UserRole.RISK_MANAGER,
}
COMMITTEE_VOTE_ROLES = COMPANY_MANAGEMENT_ROLES | {
    UserRole.BRANCH_MANAGER,
    UserRole.RISK_MANAGER,
}
FINALIZE_ROLES = COMMITTEE_VOTE_ROLES
CONDITION_UPDATE_ROLES = ANALYST_ROLES | COMMITTEE_VOTE_ROLES


class ConditionProposal(BaseModel):
    title: str = Field(min_length=2, max_length=220)
    description: str | None = Field(default=None, max_length=4000)
    condition_type: Literal["pre_contract", "pre_disbursement", "monitoring"] = "pre_disbursement"
    due_date: date | None = None


class UnderwritingAssessmentCreate(BaseModel):
    proposed_amount: Decimal | None = Field(default=None, gt=0)
    proposed_installment: Decimal | None = Field(default=None, ge=0)
    proposed_term: int | None = Field(default=None, ge=1, le=600)
    verified_income: Decimal | None = Field(default=None, ge=0)
    household_expenses: Decimal | None = Field(default=None, ge=0)
    existing_debt_installments: Decimal | None = Field(default=None, ge=0)
    dti_percent: Decimal | None = Field(default=None, ge=0)
    affordability_headroom: Decimal | None = None
    bureau_score: int | None = None
    bureau_risk_grade: str | None = Field(default=None, max_length=40)
    risk_score: Decimal | None = Field(default=None, ge=0, le=100)
    risk_grade: Literal["low", "medium", "high", "critical"]
    recommendation: Literal["approve", "approve_with_conditions", "reject", "refer"]
    rationale: str = Field(min_length=10, max_length=12000)
    strengths: list[str] = Field(default_factory=list, max_length=50)
    weaknesses: list[str] = Field(default_factory=list, max_length=50)
    exceptions: list[str] = Field(default_factory=list, max_length=50)
    mitigants: list[str] = Field(default_factory=list, max_length=50)
    proposed_conditions: list[ConditionProposal] = Field(default_factory=list, max_length=30)


class CommitteeVoteCreate(BaseModel):
    decision: Literal["approve", "approve_with_conditions", "reject", "abstain"]
    rationale: str = Field(default="", max_length=8000)
    conditions: list[ConditionProposal] = Field(default_factory=list, max_length=30)


class CommitteeFinalizeRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=8000)
    decision: Literal["approved", "conditionally_approved", "rejected"] | None = None
    override_reason: str | None = Field(default=None, max_length=8000)


class CommitteeGovernanceUpdate(BaseModel):
    required_votes: int = Field(ge=1, le=20)
    approval_threshold_percent: Decimal = Field(gt=0, le=100)
    maker_checker_required: bool = True
    reason: str = Field(default="", max_length=4000)


class ConditionUpdate(BaseModel):
    status: Literal["open", "satisfied", "waived", "failed"]
    evidence_note: str | None = Field(default=None, max_length=8000)
    waiver_reason: str | None = Field(default=None, max_length=8000)


def _case_query(db: Session, context: TenantContext):
    query = db.query(CreditCommitteeCase).filter(CreditCommitteeCase.company_id == context.company_id)
    if context.branch_id and context.role not in COMPANY_MANAGEMENT_ROLES:
        query = query.filter(CreditCommitteeCase.branch_id == context.branch_id)
    return query


@router.get("/dashboard")
def credit_committee_dashboard(
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, VIEW_ROLES)
    return dashboard_payload(db, context)


@router.get("/cases")
def list_credit_committee_cases(
    status: str | None = None,
    limit: int = Query(default=250, ge=1, le=1000),
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, VIEW_ROLES)
    query = _case_query(db, context)
    if status:
        query = query.filter(CreditCommitteeCase.status == status.strip().lower())
    rows = query.order_by(CreditCommitteeCase.created_at.desc()).limit(limit).all()
    return [case_payload(db, row) for row in rows]


@router.post("/intake/{application_id}", status_code=201)
def intake_application(
    application_id: UUID,
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, ANALYST_ROLES)
    application = application_or_404(db, context, application_id)
    if not application.credit_committee_required:
        raise HTTPException(status_code=409, detail="This historical application is not governed by the Credit Committee requirement")
    case = ensure_case(db, context, application, actor_user_id=context.user.id)
    return case_payload(db, case, include_events=True)


@router.get("/cases/{case_id}")
def credit_committee_case_detail(
    case_id: UUID,
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, VIEW_ROLES)
    return case_payload(db, case_or_404(db, context, case_id), include_events=True)


@router.patch("/cases/{case_id}/governance")
def update_committee_governance(
    case_id: UUID,
    payload: CommitteeGovernanceUpdate,
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, COMPANY_MANAGEMENT_ROLES)
    case = configure_case_governance(
        db,
        context,
        case_or_404(db, context, case_id, lock=True),
        required_votes=payload.required_votes,
        approval_threshold_percent=payload.approval_threshold_percent,
        maker_checker_required=payload.maker_checker_required,
        reason=payload.reason,
    )
    return case_payload(db, case, include_events=True)


@router.post("/cases/{case_id}/assessment", status_code=201)
def submit_underwriting_assessment(
    case_id: UUID,
    payload: UnderwritingAssessmentCreate,
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, ANALYST_ROLES)
    row = submit_assessment(
        db,
        context,
        case_or_404(db, context, case_id, lock=True),
        proposed_amount=payload.proposed_amount,
        proposed_installment=payload.proposed_installment,
        proposed_term=payload.proposed_term,
        verified_income=payload.verified_income,
        household_expenses=payload.household_expenses,
        existing_debt_installments=payload.existing_debt_installments,
        dti_percent=payload.dti_percent,
        affordability_headroom=payload.affordability_headroom,
        bureau_score=payload.bureau_score,
        bureau_risk_grade=payload.bureau_risk_grade,
        risk_score=payload.risk_score,
        risk_grade=payload.risk_grade,
        recommendation=payload.recommendation,
        rationale=payload.rationale,
        strengths=payload.strengths,
        weaknesses=payload.weaknesses,
        exceptions=payload.exceptions,
        mitigants=payload.mitigants,
        proposed_conditions=[item.model_dump(mode="json") for item in payload.proposed_conditions],
    )
    return assessment_payload(row)


@router.put("/cases/{case_id}/vote")
def record_committee_vote(
    case_id: UUID,
    payload: CommitteeVoteCreate,
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, COMMITTEE_VOTE_ROLES)
    row = cast_vote(
        db,
        context,
        case_or_404(db, context, case_id, lock=True),
        decision=payload.decision,
        rationale=payload.rationale,
        conditions=[item.model_dump(mode="json") for item in payload.conditions],
    )
    case = case_or_404(db, context, case_id)
    return {"vote": {"id": str(row.id), "decision": row.decision}, "case": case_payload(db, case, include_events=True)}


@router.post("/cases/{case_id}/finalize")
def finalize_committee_case(
    case_id: UUID,
    payload: CommitteeFinalizeRequest,
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, FINALIZE_ROLES)
    case = finalize_case(
        db,
        context,
        case_or_404(db, context, case_id, lock=True),
        reason=payload.reason,
        requested_decision=payload.decision,
        override_reason=payload.override_reason,
        allow_override=context.role in COMPANY_MANAGEMENT_ROLES,
    )
    return case_payload(db, case, include_events=True)


@router.patch("/cases/{case_id}/conditions/{condition_id}")
def resolve_committee_condition(
    case_id: UUID,
    condition_id: UUID,
    payload: ConditionUpdate,
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, CONDITION_UPDATE_ROLES)
    if payload.status == "waived" and context.role not in (COMPANY_MANAGEMENT_ROLES | {UserRole.RISK_MANAGER}):
        raise HTTPException(status_code=403, detail="Only company management or the risk manager may waive a credit condition")
    row = update_condition(
        db,
        context,
        case_or_404(db, context, case_id, lock=True),
        condition_id,
        status=payload.status,
        evidence_note=payload.evidence_note,
        waiver_reason=payload.waiver_reason,
    )
    case = case_or_404(db, context, case_id)
    return {"condition": condition_payload(row), "case": case_payload(db, case, include_events=True)}


@router.get("/cases/{case_id}/credit-memo.pdf")
def credit_committee_pdf(
    case_id: UUID,
    db: Session = Depends(get_db),
    context: TenantContext = Depends(get_user_context),
):
    require_tenant_roles(context, VIEW_ROLES)
    case = case_or_404(db, context, case_id)
    content = build_credit_committee_pdf(db, case)
    return Response(
        content=content,
        media_type="application/pdf",
        headers={"Content-Disposition": f"attachment; filename={case.case_reference}-credit-memo.pdf"},
    )

File: apps/frontend/api/creditCommittee.ts
import { api } from "@/lib/api";

export type CreditCondition = {
  id: string;
  source: string;
  condition_type: "pre_contract" | "pre_disbursement" | "monitoring";
  title: string;
  description: string | null;
  status: "open" | "satisfied" | "waived" | "failed";
  due_date: string | null;
  evidence_note: string | null;
  waiver_reason: string | null;
};

export type CommitteeVote = {
  id: string;
  user_id: string;
  role: string;
  decision: "approve" | "approve_with_conditions" | "reject" | "abstain";
  rationale: string;
  conditions: Array<{ title: string; description?: string | null; condition_type?: string; due_date?: string | null }>;
  voted_at: string | null;
};

export type UnderwritingAssessment = {
  id: string;
  revision: number;
  status: string;
  analyst_user_id: string | null;
  requested_amount: number;
  proposed_amount: number;
  proposed_installment: number;
  proposed_term: number;
  verified_income: number;
  household_expenses: number;
  existing_debt_installments: number;
  dti_percent: number;
  affordability_headroom: number;
  bureau_score: number | null;
  bureau_risk_grade: string | null;
  kyc_status: string | null;
  risk_score: number | null;
  risk_grade: string;
  recommendation: string;
  rationale: string;
  strengths: string[];
  weaknesses: string[];
  exceptions: string[];
  mitigants: string[];
  proposed_conditions: Array<{ title: string; description?: string | null; condition_type?: string; due_date?: string | null }>;
  evidence_snapshot: Record<string, unknown>;
  submitted_at: string | null;
};

export type CreditCommitteeCase = {
  id: string;
  case_reference: string;
  application_id: string;
  application_reference: string | null;
  application_status: string | null;
  borrower_id: string;
  borrower_name: string;
  requested_amount: number;
  term_count: number;
  application_type: string | null;
  channel: string | null;
  status: string;
  required_votes: number;
  approval_threshold_percent: number;
  maker_checker_required: boolean;
  analyst_user_id: string | null;
  analyst_submitted_at: string | null;
  evidence_snapshot: Record<string, any>;
  latest_assessment: UnderwritingAssessment | null;
  votes: CommitteeVote[];
  vote_summary: {
    counts: Record<string, number>;
    decisive_vote_count: number;
    approve_like_count: number;
    approval_ratio_percent: number;
    required_votes: number;
    approval_threshold_percent: number;
    quorum_met: boolean;
    computed_outcome: string | null;
  };
  conditions: CreditCondition[];
  final_decision: string | null;
  final_decision_reason: string | null;
  final_decided_at: string | null;
  override_used: boolean;
  override_reason: string | null;
  locked_at: string | null;
  created_at: string | null;
  events?: Array<{ id: string; event_type: string; actor_user_id: string | null; payload: Record<string, unknown>; created_at: string | null }>;
};

export type CommitteeDashboard = {
  summary: {
    awaiting_intake: number;
    underwriting: number;
    committee_review: number;
    awaiting_conditions: number;
    approved: number;
    rejected: number;
  };
  awaiting_intake: Array<{
    application_id: string;
    application_reference: string;
    borrower_id: string;
    borrower_name: string;
    requested_amount: number;
    term_count: number;
    status: string;
    submitted_at: string | null;
  }>;
  cases: CreditCommitteeCase[];
};

export type ConditionProposal = {
  title: string;
  description?: string | null;
  condition_type?: "pre_contract" | "pre_disbursement" | "monitoring";
  due_date?: string | null;
};

export async function getCommitteeDashboard(): Promise<CommitteeDashboard> {
  return (await api.get<CommitteeDashboard>("/credit-committee/dashboard")).data;
}

export async function intakeCommitteeApplication(applicationId: string): Promise<CreditCommitteeCase> {
  return (await api.post<CreditCommitteeCase>(`/credit-committee/intake/${applicationId}`)).data;
}

export async function getCommitteeCase(caseId: string): Promise<CreditCommitteeCase> {
  return (await api.get<CreditCommitteeCase>(`/credit-committee/cases/${caseId}`)).data;
}

export async function submitUnderwritingAssessment(caseId: string, payload: {
  proposed_amount?: number;
  proposed_installment?: number;
  proposed_term?: number;
  verified_income?: number;
  household_expenses?: number;
  existing_debt_installments?: number;
  dti_percent?: number;
  affordability_headroom?: number;
  bureau_score?: number;
  bureau_risk_grade?: string;
  risk_score?: number;
  risk_grade: "low" | "medium" | "high" | "critical";
  recommendation: "approve" | "approve_with_conditions" | "reject" | "refer";
  rationale: string;
  strengths: string[];
  weaknesses: string[];
  exceptions: string[];
  mitigants: string[];
  proposed_conditions: ConditionProposal[];
}): Promise<UnderwritingAssessment> {
  return (await api.post<UnderwritingAssessment>(`/credit-committee/cases/${caseId}/assessment`, payload)).data;
}

export async function castCommitteeVote(caseId: string, payload: {
  decision: "approve" | "approve_with_conditions" | "reject" | "abstain";
  rationale: string;
  conditions: ConditionProposal[];
}) {
  return (await api.put<{ vote: { id: string; decision: string }; case: CreditCommitteeCase }>(`/credit-committee/cases/${caseId}/vote`, payload)).data;
}

export async function finalizeCommitteeCase(caseId: string, payload: {
  reason?: string;
  decision?: "approved" | "conditionally_approved" | "rejected";
  override_reason?: string;
}) {
  return (await api.post<CreditCommitteeCase>(`/credit-committee/cases/${caseId}/finalize`, payload)).data;
}

export async function updateCommitteeCondition(caseId: string, conditionId: string, payload: {
  status: "open" | "satisfied" | "waived" | "failed";
  evidence_note?: string;
  waiver_reason?: string;
}) {
  return (await api.patch<{ condition: CreditCondition; case: CreditCommitteeCase }>(`/credit-committee/cases/${caseId}/conditions/${conditionId}`, payload)).data;
}

export async function updateCommitteeGovernance(caseId: string, payload: {
  required_votes: number;
  approval_threshold_percent: number;
  maker_checker_required: boolean;
  reason: string;
}) {
  return (await api.patch<CreditCommitteeCase>(`/credit-committee/cases/${caseId}/governance`, payload)).data;
}

export async function downloadCreditMemo(item: CreditCommitteeCase) {
  const response = await api.get<Blob>(`/credit-committee/cases/${item.id}/credit-memo.pdf`, { responseType: "blob" });
  const url = URL.createObjectURL(response.data);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = `${item.case_reference}-credit-memo.pdf`;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  window.setTimeout(() => URL.revokeObjectURL(url), 5_000);
}
